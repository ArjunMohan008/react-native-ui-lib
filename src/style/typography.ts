import _ from 'lodash';
import {Platform} from 'react-native';
import Constants from '../helpers/Constants';

import TypographyPresets from './typographyPresets';

export const WEIGHT_TYPES = { //do we need this? we have a duplicate object inside the presets
  THIN: '200',
  LIGHT: '300',
  REGULAR: '400',
  MEDIUM: parseFloat(Platform.Version as string) >= 11.2 ? '600' : '500',
  BOLD: '700',
  HEAVY: '800',
  BLACK: '900'
};

export class Typography {
  keysPattern = this.generateKeysPattern();

  /**
   * Load custom set of typographies
   * arguments:
   * typographies - map of keys and typography values
   * e.g {text15: {fontSize: 58, fontWeight: '100', lineHeight: Math.floor(58 * 1.4)}}
   */
  loadTypographies(typographies: Dictionary<any>) {
    _.forEach(typographies, (value, key) => {
      //@ts-ignore
      this[key] = value;
    });
    this.keysPattern = this.generateKeysPattern();
  }

  getKeysPattern() {
    return this.keysPattern;
  }

  generateKeysPattern() {
    return new RegExp(_.chain(this)
      .keys()
      .map(key => [`${key}`])
      .flatten()
      .join('|')
      .value());
  }

  // TODO: deprecate
  async measureWidth(text: string, typography = TypographyPresets.text70, containerWidth = Constants.screenWidth) {
    const size = await this.measureTextSize(text, typography, containerWidth);
    if (size) {
      return size.width;
    }
  }

  async measureTextSize(text: string, typography = TypographyPresets.text70, containerWidth = Constants.screenWidth) {
    const rnTextSize = require('react-native-text-size').default;
    if (text) {
      const size = await rnTextSize.measure({
        text, // text to measure, can include symbols
        width: containerWidth, // max-width of the "virtual" container
        ...typography, // RN font specification
      });
      return size;
    }
  }
}
const TypedTypography = Typography as ExtendTypeWith<typeof Typography, typeof TypographyPresets>
const typography = new TypedTypography();
typography.loadTypographies(TypographyPresets);

export default typography;
