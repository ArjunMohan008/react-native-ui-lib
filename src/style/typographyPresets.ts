import {Platform, TextStyle} from 'react-native';
import _ from 'lodash';
import Constants from '../helpers/Constants';

export const WEIGHT_TYPES: {[key: string]: TextStyle['fontWeight']} = {
  THIN: '200' as '200',
  LIGHT: '300' as '300',
  REGULAR: '400' as '400',
  MEDIUM:
    parseFloat(Platform.Version as string) >= 11.2
      ? '600'
      : ('500' as '500' | '600'),
  BOLD: '700' as '700',
  HEAVY: '800' as '800',
  BLACK: '900' as '900'
};

// text10
const text10: TextStyle = {
  fontSize: 64,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.THIN : undefined,
  lineHeight: 76,
  fontFamily: 'System'
};

// text20
const text20: TextStyle = {
  fontSize: 48,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.REGULAR : undefined,
  lineHeight: Constants.isIOS ? 60 : 62,
  fontFamily: 'System'
};

// text30
const text30: TextStyle = {
  fontSize: 36,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.REGULAR : undefined,
  lineHeight: Constants.isIOS ? 43 : 46,
  fontFamily: 'System'
};

// text40
const text40: TextStyle = {
  fontSize: 28,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.HEAVY : undefined,
  lineHeight: 32,
  fontFamily: 'System'
};

// text50
const text50: TextStyle = {
  fontSize: 24,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.HEAVY : undefined,
  lineHeight: 28,
  fontFamily: 'System'
};

// text60
const text60: TextStyle = {
  fontSize: 20,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.HEAVY : undefined,
  lineHeight: 24,
  fontFamily: 'System'
};

// text65
const text65: TextStyle = {
  fontSize: 18,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.MEDIUM : undefined,
  lineHeight: 24,
  fontFamily: 'System'
};

// text70
const text70: TextStyle = {
  fontSize: 16,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.REGULAR : undefined,
  lineHeight: 24,
  fontFamily: 'System'
};

// text80
const text80: TextStyle = {
  fontSize: 14,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.REGULAR : undefined,
  lineHeight: 20,
  fontFamily: 'System'
};

// text90
const text90: TextStyle = {
  fontSize: 12,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.BOLD : undefined,
  lineHeight: 16,
  fontFamily: 'System'
};

// text100
const text100: TextStyle = {
  fontSize: 10,
  fontWeight: Constants.isIOS ? WEIGHT_TYPES.BOLD : undefined,
  lineHeight: 16,
  fontFamily: 'System'
};

const Typography: {[key: string]: TextStyle} = {
  text10,
  text20,
  text30,
  text40,
  text50,
  text60,
  text65,
  text70,
  text80,
  text90,
  text100
};

const keys = [10, 20, 30, 40, 50, 60, 65, 70, 80, 90, 100];
const weightsMap = {
  THIN: 'T',
  LIGHT: 'L',
  REGULAR: 'R',
  MEDIUM: 'M',
  BOLD: 'BO',
  HEAVY: 'H',
  BLACK: 'BL'
};

_.forEach(keys, (key) => {
  _.forEach(weightsMap, (weightValue, weightKey) => {
    const fontKey = `text${key}`;
    const fontWeightKey = `${fontKey}${weightValue}`;
    Typography[fontWeightKey] = {
      ...Typography[fontKey],
      fontWeight: Constants.isIOS ? WEIGHT_TYPES[weightKey] : undefined
    };
  });
});

export default Typography;
