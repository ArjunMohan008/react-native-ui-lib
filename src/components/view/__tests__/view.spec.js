import React from 'react';
import renderer from 'react-test-renderer';
import View from '../index';
import {ThemeManager} from '../../../style';

describe('View', () => {
  beforeEach(() => {
    ThemeManager.setComponentTheme('View', undefined);
  });

  it('should render clean view', () => {
    const tree = renderer.create(<View />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('should support theme props', () => {
    it('should have default theme props', () => {
      ThemeManager.setComponentTheme('View', {
        'bg-dark80': true,
        flex: true,
      });
      const tree = renderer.create(<View />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('layout modifiers', () => {
    it('should render view with flex modifier', () => {
      const tree = renderer.create(<View flex />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with row modifier', () => {
      const tree = renderer.create(<View row />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with left+bottom modifier', () => {
      const tree = renderer.create(<View left bottom />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with right+top modifier', () => {
      const tree = renderer.create(<View right top />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('style modifiers', () => {
    it('should render view with red50 background color modifier', () => {
      const tree = renderer.create(<View bg-red50 />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with black background color modifier', () => {
      const tree = renderer.create(<View bg-black />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with border radius modifier', () => {
      const tree = renderer.create(<View br20 />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('spacings modifiers', () => {
    // Paddings
    it('should render view with right and left padding', () => {
      const tree = renderer.create(<View paddingR-20 paddingL-10 />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with top and bottom padding', () => {
      const tree = renderer.create(<View paddingR-20 paddingL-10 />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with horizontal padding', () => {
      const tree = renderer.create(<View paddingH-20 />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with vertical padding', () => {
      const tree = renderer.create(<View paddingV-20 />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    // Margins
    it('should render view with right and left margin', () => {
      const tree = renderer.create(<View marginR-20 marginL-10 />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with top and bottom margin', () => {
      const tree = renderer.create(<View marginR-20 marginL-10 />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with horizontal margin', () => {
      const tree = renderer.create(<View marginH-20 />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render view with vertical margin', () => {
      const tree = renderer.create(<View marginV-20 />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
