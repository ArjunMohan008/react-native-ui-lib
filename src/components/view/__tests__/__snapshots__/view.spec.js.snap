// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View layout modifiers should render view with flex modifier 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
    ]
  }
/>
`;

exports[`View layout modifiers should render view with left+bottom modifier 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "flex-start",
        "justifyContent": "flex-end",
      },
    ]
  }
/>
`;

exports[`View layout modifiers should render view with right+top modifier 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "flex-end",
        "justifyContent": "flex-start",
      },
    ]
  }
/>
`;

exports[`View layout modifiers should render view with row modifier 1`] = `
<View
  row={true}
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
    ]
  }
/>
`;

exports[`View should render clean view 1`] = `
<View
  style={Array []}
/>
`;

exports[`View spacings modifiers should render view with horizontal margin 1`] = `
<View
  marginH-20={true}
  style={
    Array [
      Object {
        "marginHorizontal": 20,
      },
    ]
  }
/>
`;

exports[`View spacings modifiers should render view with horizontal padding 1`] = `
<View
  paddingH-20={true}
  style={
    Array [
      Object {
        "paddingHorizontal": 20,
      },
    ]
  }
/>
`;

exports[`View spacings modifiers should render view with right and left margin 1`] = `
<View
  marginL-10={true}
  marginR-20={true}
  style={
    Array [
      Object {
        "marginLeft": 10,
        "marginRight": 20,
      },
    ]
  }
/>
`;

exports[`View spacings modifiers should render view with right and left padding 1`] = `
<View
  paddingL-10={true}
  paddingR-20={true}
  style={
    Array [
      Object {
        "paddingLeft": 10,
        "paddingRight": 20,
      },
    ]
  }
/>
`;

exports[`View spacings modifiers should render view with top and bottom margin 1`] = `
<View
  marginL-10={true}
  marginR-20={true}
  style={
    Array [
      Object {
        "marginLeft": 10,
        "marginRight": 20,
      },
    ]
  }
/>
`;

exports[`View spacings modifiers should render view with top and bottom padding 1`] = `
<View
  paddingL-10={true}
  paddingR-20={true}
  style={
    Array [
      Object {
        "paddingLeft": 10,
        "paddingRight": 20,
      },
    ]
  }
/>
`;

exports[`View spacings modifiers should render view with vertical margin 1`] = `
<View
  marginV-20={true}
  style={
    Array [
      Object {
        "marginVertical": 20,
      },
    ]
  }
/>
`;

exports[`View spacings modifiers should render view with vertical padding 1`] = `
<View
  paddingV-20={true}
  style={
    Array [
      Object {
        "paddingVertical": 20,
      },
    ]
  }
/>
`;

exports[`View style modifiers should render view with black background color modifier 1`] = `
<View
  bg-black={true}
  style={
    Array [
      Object {
        "backgroundColor": "#000000",
      },
    ]
  }
/>
`;

exports[`View style modifiers should render view with border radius modifier 1`] = `
<View
  br20={true}
  style={
    Array [
      Object {
        "borderRadius": 6,
      },
    ]
  }
/>
`;

exports[`View style modifiers should render view with red50 background color modifier 1`] = `
<View
  bg-red50={true}
  style={
    Array [
      Object {
        "backgroundColor": "#F79A94",
      },
    ]
  }
/>
`;
